builtin/providers/heroku/resource_heroku_formation.go

```diff
@@ -0,0 +1,116 @@
+package heroku
+
+import (
+	"fmt"
+
+	"github.com/cyberdelia/heroku-go/v3"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceHerokuFormation() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceHerokuFormationCreate,
+		Read:   resourceHerokuFormationRead,
+		Update: resourceHerokuFormationUpdate,
+		Delete: resourceHerokuFormationDelete,
+		Exists: resourceHerokuFormationExists,
+
+		Schema: map[string]*schema.Schema{
+			"app": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"type": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"quantity": &schema.Schema{
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+
+			"size": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+func resourceHerokuFormationCreate(d *schema.ResourceData, meta interface{}) error {
+
+	d.SetId("")
+	return resourceHerokuFormationUpdate(d, meta)
+}
+
+func resourceHerokuFormationRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*heroku.Service)
+
+	formations, err := client.FormationList(d.Get("app").(string), (&heroku.ListRange{Max: 10000}))
+
+	if err != nil {
+		return fmt.Errorf("Error retrieving formation list: %s", err)
+	}
+
+	for _, formation := range formations {
+		if formation.Type == d.Get("type").(string) {
+			d.Set("size", formation.Size)
+			d.Set("quantity", formation.Quantity)
+			d.SetId(formation.ID)
+		}
+	}
+
+	return nil
+}
+
+func resourceHerokuFormationUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*heroku.Service)
+
+	app := d.Get("app").(string)
+
+	quantity := d.Get("quantity").(int)
+	size := d.Get("size").(string)
+
+	if d.Id() == d.Get("type") && (d.HasChange("quantity") || d.HasChange("size")) {
```

Comment:
following that loop though, if `d.SetId()` gets called with the formation id from Heroku in the `read` method, when then will `d.Id() == d.Get("type")` evaluate to `true`?